docs/src/design/chiplets/kernel_rom.md

```diff
@@ -1,91 +1,101 @@
 $$
 
-These constraints on $idx$ should not be applied to the very last row of the kernel ROM's execution trace, since we do not want to enforce a value that would conflict with the first row of a subsequent chiplet (or padding). Therefore we can create a special virtual flag for this constraint using the $chip\_s_3$ selector column from the [chiplets](main.md) module that selects for the kernel ROM chiplet.
+_Note that this selector need not be multiplied by the kernel ROM chiplet flag $chip\_s_4$, since it is only active when the previous chiplet is active._
 
-The modified constraints which should be applied are the following:
+The hash of the procedure repeats in the next row whenever the next row is not the first of a new procedure, i.e., when $s_{first}' = 0$.
+To ensure the constraint is disabled in the last row, we multiply it by the chiplet selector $s_4$ for the next row,
+since it transitions from 0 to 1 at this point, indicating the start of
 
->$$
-(1 - chip\_s_3') \cdot \Delta idx \cdot (1 - \Delta idx) = 0 \text{ | degree} = 3
+> $$
+s_4' \cdot (1 - s_{first}') \cdot (r_i' - r_i) = 0 \text{ | degree} = 3
 $$
 
->$$
-(1 - chip\_s_3') \cdot (1 - \Delta idx) \cdot (r_i' - r_i) = 0 \text{ | degree} = 3
-$$
+### Kernel procedure table constraints
 
-_Note: these constraints should also be multiplied by chiplets module's selector flag for the kernel ROM chiplet, as is true for all constraints in this chiplet._
+This kernel procedure table keeps track of all *unique* kernel function roots. The values in this table will be updated
+only when the value in the $s_{first}$ selector is set.
 
-## Chiplets bus constraints
+The chiplet communicates with the chiplet bus via the $b_{chip}$ columns.
+In the first row of each trace segment containing the same hash, it responds to a request with message $u_{init}$,
+ensuring equality between the set of hashes contained in the trace and the list of kernel procedure roots provided via
+public inputs.
+In all other rows, the chiplet responds messages $u_{bus}$ requested during program block hashing of the [
+`SYSCALL` operation](../decoder/constraints.md#block-hash-computation-constraints).
 
-The chiplets bus is used to keep track of all kernel function calls. To simplify the notation for describing kernel ROM constraints on the chiplets bus, we'll first define variable $u$, which represents how each kernel procedure in the kernel ROM's execution trace is reduced to a single value. Denoting the random values received from the verifier as $\alpha_0, \alpha_1$, etc., this can be achieved as follows.
+The variables $u_{init}$ and $u_{bus}$ represent reduced bus messages containing a kernel procedure hash.
+Denoting the random values received from the verifier as $\alpha_0, \alpha_1$, etc., this can be defined as
 
 $$
-v = \alpha_0 + \alpha_1 \cdot op_{krom} + \sum_{i=0}^3 (\alpha_{i + 2} \cdot r_i)
+\begin{aligned}
+u_{init} &= \alpha_0 + \sum_{i=0}^3 (\alpha_{i + 1} \cdot r_i) \\
```

Comment:
As you mentioned, we should domain separate. Using $0$ as the value of the domain separator for this message will simplify things a bit to

$$u_{init} = \alpha_0 + \sum_{i=0}^3 (\alpha_{i + 2} \cdot r_i)$$

But any separator will be equal from a recursive verifier cost point of view.

If it is not much work, let's do it in this PR

Reply:
Are you suggesting that it's not more a hassle to use a non-zero label for the domain separator?

Reply:
Yes, if we are using the following message hashing method $$u_{init} = \alpha + \textsf{label} \cdot \beta^0 + \sum_{i=0}^3 (\beta^{i + 1} \cdot r_i)$$, then the value of $\textsf{label}$ is not important from verifier-work point of view.